### DISCUSSION TITLE
Cluster vectors for very high dimensional graphs

### DISCUSSION THREAD

**Discussion by hkim716** [pyg-team/pytorch_geometric#2830] [Q&A] [answered]
Hi Matt, thanks for very helpful comments before. 
I'm trying to use different dataset, which has tens of thousands `nodes` and `edges`. 
This dataset also has fixed geometry with 3-dimensional `pos` values, `x`, `y`, `z`.
Each example looks like `Data(edge_attr=[1459862, 3], edge_index=[2, 1459862], pos=[366883, 3], x=[366883, 1])`, and I have 100 examples as entire dataset. (`edge_attr`, `edge_index`, and `pos` are the same for all examples but `x`(node feature), as my previous discussion)
I tried to obtain `cluster` vectors for a single example using

``` python
def pre_process(train_dataset, depth):
    train_loader_pre = DataLoader(train_dataset, batch_size=1, shuffle=False)
    clusters = []
    for data in train_loader_pre:
        for i in range(depth):
            x, edge_index, edge_attr, pos = data.x, data.edge_index, data.edge_attr, data.pos
            cluster = graclus(edge_index)
            cluster, _ = consecutive_cluster(cluster)
            clusters.append(cluster)
            data = max_pool(clusters[i], data, transform=transform)
        break
 
    return clusters
```
However, when I tried to run `pre_process` function, there was an error caused with `depth > 5`.
Sometimes I can obtain `cluster` vectors with `depth=6` (succeeded once when I tried ten times), but most of the time, it's hard to get them with large value of `depth`.
I think it's because my dataset has too many `nodes` and `edges` to cluster, but is it common issue? or is there better way to do it?

**Comment by rusty1s** [comment, marked as answer]
I think the `graclus` method currently has problems dealing with isolated nodes. Can you check if there exists isolated nodes in your graph after a certain amount of coarsening?

**Reply by hkim716** [reply]
``` python
----> 2 clusters = pre_process(train_dataset, depth=depth)

<ipython-input-4-5a5899ce67b4> in pre_process(train_dataset, depth)
     10             cluster, _ = consecutive_cluster(cluster)
     11             clusters.append(cluster)
---> 12             data = max_pool(clusters[i], data, transform=transform)
     13         break
     14 

~/miniconda3/envs/gnb_df_pyg/lib/python3.7/site-packages/torch_geometric/nn/pool/max_pool.py in max_pool(cluster, data, transform)
     66     cluster, perm = consecutive_cluster(cluster)
     67 
---> 68     x = None if data.x is None else _max_pool_x(cluster, data.x)
     69     index, attr = pool_edge(cluster, data.edge_index, data.edge_attr)
     70     batch = None if data.batch is None else pool_batch(perm, data.batch)

~/miniconda3/envs/gnb_df_pyg/lib/python3.7/site-packages/torch_geometric/nn/pool/max_pool.py in _max_pool_x(cluster, x, size)
     10 
     11 def _max_pool_x(cluster, x, size: Optional[int] = None):
---> 12     return scatter(x, cluster, dim=0, dim_size=size, reduce='max')
     13 
     14 

~/miniconda3/envs/gnb_df_pyg/lib/python3.7/site-packages/torch_scatter/scatter.py in scatter(src, index, dim, out, dim_size, reduce)
    156         return scatter_min(src, index, dim, out, dim_size)[0]
    157     elif reduce == 'max':
--> 158         return scatter_max(src, index, dim, out, dim_size)[0]
    159     else:
    160         raise ValueError

RuntimeError: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript (most recent call last):
  File "/home/hkim/miniconda3/envs/gnb_df_pyg/lib/python3.7/site-packages/torch_scatter/scatter.py", line 72, in scatter_max
        out: Optional[torch.Tensor] = None,
        dim_size: Optional[int] = None) -> Tuple[torch.Tensor, torch.Tensor]:
    return torch.ops.torch_scatter.scatter_max(src, index, dim, out, dim_size)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
RuntimeError: The expanded size of the tensor (209338) must match the existing size (209337) at non-singleton dimension 0.  Target sizes: [209338, 1].  Tensor sizes: [209337, 1]
```

I think you are right because `max_pool` got an error when it comes to get different size of tensors. Does the difference between `(209338)` and `(209337)`, which is `1`, mean the number of isolated nodes? I think it happens after 4-coarsening steps.

**Reply by rusty1s** [reply]
Yeah, it looks like `x` is holding one node less than what is expected by `cluster`, which suggests that there is an isolated node that does not get properly tracked by `graclus`.

Can you run
```python
assert edge_index.unique().numel() == x.size(0) 
```
before calling `graclus` to confirm?

**Reply by hkim716** [reply]
Yes, that was the problem. When I ran 

``` python
count = 0
for i in range(depth):
    count+=1
    x, edge_index, edge_attr, pos = data.x, data.edge_index, data.edge_attr, data.pos
    print(count)
    cluster = graclus(edge_index)
    cluster, _ = consecutive_cluster(cluster)
    assert cluster.size(0) == x.size(0)
    clusters.append(cluster)
    data = max_pool(clusters[i], data, transform=transform)
break
```
I got these outputs
```
1
2
3
4
5
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
---> 13             assert cluster.size(0) == x.size(0)
```

Then I tried 
``` python
stopper = False
while stopper == False:
    cluster = graclus(edge_index)
    cluster, _ = consecutive_cluster(cluster)
    if cluster.size(0) == x.size(0):
        stopper = True
```
``` python
clusters = pre_process(train_dataset, depth=5)
```
But the process does not end because `cluster.size(0) == x.size(0)` was not satisfied for the 5th `cluster`.
Sometimes `depth=5` works, but I've never get `[clusters]` with `depth=6`, which means there is always at least one isolated node after 5th `cluster`. Would you have any suggestion in that case?

**Reply by rusty1s** [reply]
Just found a way to fix it: 
```python
cluster = graclus(edge_index, num_nodes=x.size(0))
```

**Reply by hkim716** [reply]
Thanks Matt, that solved the problem! 
Now I have another issue that the size of `cluster` vector does not decrease anymore after 19th `[clusters]`.
The size of `cluster` vector is like this. 
`[366883]->[280705]->[234120]->[209265]->[196075]->[189142]->....->[181601]->[181598]->[181597]->[181596]->[181596]->[181596]->[181596]` 
As you can see, the size of `cluster` is gradually converging to `181596`. I think that's because as step goes by, `max_pool` does not want to define new neighborhoods. I tried with another `data` example in `train_loader_pre`, which has different `x_feature` values, but they always went to `[181596]`. How could I generate `cluster` vectors that always decrease in size?

**Reply by rusty1s** [reply]
The size or content of `cluster` is solely determined by the graph structure, and is independent of node features values. In general, `graclus` tries to halve the number of nodes on each execution. The only case in which `graclus` fails to do this is in case a node does not have a neighbor with whom it can group to a new "super" node. 

You can also send me the `edge_index` and let me take a look.

**Reply by hkim716** [reply]
Thanks for your help Matt. I found some self-connected nodes on my original graph. There were 2432 self-connected nodes. However, even after I deleted them in my original dataset, the smallest size of `cluster` vector I can get is still `[181596]`.
Here's my `edge_index`, which is the version without self-connected nodes, such as `[285, 285]`.

[edge_index.zip](https://github.com/rusty1s/pytorch_geometric/files/6809191/edge_index.zip)

**Reply by rusty1s** [reply]
The reason why `graclus` cannot further reduce the number of nodes is because there exists exactly `181590` isolated nodes in graph:
```python
from torch_geometric.utils import degree
deg = degree(edge_index[0])
print((deg == 0).sum())
```
You can probably those in a pre-processing step:
```python
edge_index, _ = subgraph(deg > 0, edge_index, relabel_nodes=True)
```

**Reply by hkim716** [reply]
Thank you so much Matt, that solved the problem!
In that case, is there any difference between when I'm doing it in `pre-process` function and doing this when I create original `edge_index` using `MyDataset.py` to get rid of all of isolated nodes?

``` python
class MyData(InMemoryDataset):
    ...

    def process(self):
        raw_data = np.load(self.raw_paths[0])

        x = torch.from_numpy(raw_data['x'])
        z = torch.from_numpy(raw_data['z'])

        pos = torch.stack([x, z], dim=-1)
        value = torch.from_numpy(raw_data['value'])
        y = torch.from_numpy(raw_data['label'])

        edge_index = torch.from_numpy(raw_data['edge_index'])
        edge_index -= 1
        edge_index = edge_index.t().contiguous()
        edge_index = to_undirected(edge_index)

        data_list = []
        for i in range(cp.size(0)):
            data = Data(x=value[i].view(-1,1),pos=pos.view(-1,2), edge_index=edge_index, y=y[i,:].view(-1))
            data_list.append(data)

        torch.save(self.collate(data_list), self.processed_paths[0])
```

**Reply by rusty1s** [reply]
There shouldn't be a difference. I suggest to do these kinds of transformations as a pre-processing step though to disentangle data representation and model.

**Reply by hkim716** [reply]
Thank you so much Matt. I really appreciate your help! I'm getting interested in this GNN world and hope I can also contribute on this awesome library later!
